graph/src/ext/futures.rs

```diff
@@ -0,0 +1,123 @@
+        if self.canceled.poll().is_err() {
+            Err((self.on_cancel)())
+        } else {
+            self.cancelable.poll()
+        }
+    }
+}
+
+impl<F: Future, C: Fn() -> F::Error> Future for Cancelable<F, C> {
+    type Item = F::Item;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        // Error if the future was canceled, otherwise poll it.
+        if self.canceled.poll().is_err() {
+            Err((self.on_cancel)())
+        } else {
+            self.cancelable.poll()
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct CancelGuard {
+    canceler: oneshot::Sender<()>,
+}
+
+impl CancelGuard {
+    /// A more readable `drop`.
+    pub fn cancel(self) {}
+
+    /// Convert into a `SharedGuard`.
+    pub fn shared(self) -> SharedCancelGuard {
+        SharedCancelGuard {
+            guard: Arc::new(Mutex::new(Some(self.canceler))),
+        }
+    }
+}
+
+/// A version of `CancelGuard` that can be cloned.
+///
+/// To cancel the stream, call `cancel` on one of the guards or drop all guards.
+#[derive(Clone, Debug)]
+pub struct SharedCancelGuard {
+    guard: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+}
+
+impl SharedCancelGuard {
+    /// Cancels the stream, a noop if already canceled.
+    pub fn cancel(&self) {
+        *self.guard.lock().unwrap() = None
+    }
+
+    pub fn is_canceled(&self) -> bool {
+        self.guard.lock().unwrap().is_none()
+    }
+}
+
+pub trait StreamExtension: Stream + Sized {
+    /// When `cancel` is called on a `CancelGuard` or it is dropped,
+    /// `Cancelable` receives an error.
+    ///
+    /// `on_cancel` is called to make an error value upon cancelation.
+    fn cancelable<C: Fn() -> Self::Error>(self, on_cancel: C)
+        -> (Cancelable<Self, C>, CancelGuard);
+}
+
+impl<S: Stream> StreamExtension for S {
+    fn cancelable<C: Fn() -> S::Error>(self, on_cancel: C) -> (Cancelable<Self, C>, CancelGuard) {
+        let (canceler, canceled) = oneshot::channel();
+        (
+            Cancelable {
+                cancelable: self,
+                canceled,
+                on_cancel,
+            },
+            CancelGuard { canceler },
+        )
+    }
+}
+
+pub trait FutureExtension: Future + Sized {
+    /// When `cancel` is called on a `CancelGuard` or it is dropped,
+    /// `Cancelable` receives an error.
+    ///
+    /// `on_cancel` is called to make an error value upon cancelation.
+    fn cancelable<C: Fn() -> Self::Error>(self, on_cancel: C)
+        -> (Cancelable<Self, C>, CancelGuard);
+}
+
+impl<F: Future> FutureExtension for F {
+    fn cancelable<C: Fn() -> F::Error>(self, on_cancel: C) -> (Cancelable<Self, C>, CancelGuard) {
+        let (canceler, canceled) = oneshot::channel();
+        (
+            Cancelable {
+                cancelable: self,
+                canceled,
+                on_cancel,
+            },
+            CancelGuard { canceler },
+        )
+    }
```

Comment:
I think we need a way to cancel a bunch of futures/streams with one canceler. Is that possible with this design?

Reply:
It's not currently possible, this design can be extended to allow that, I can do it in a follow-up PR.

Reply:
Yeah, I think it'd be worth iterating on this design -- we're going to be using this everywhere

Reply:
I'm thinking it'd be a matter of adding two methods, one to create the a `SharedCancelGuard` along with a shared receiver, then a method `cancelable_with` that takes a shared receiver.